pkg/collect/host_filesystem_performance_linux.go

```diff
@@ -143,83 +181,116 @@ func collectHostFilesystemPerformance(hostCollector *troubleshootv1beta2.Filesys
+	for i := 0; i < opts.jobs; i++ {
 		wg.Add(1)
 
 		go func(i int) {
-			defer wg.Done()
-
-			data := make([]byte, int(operationSize))
-			fd := int(f.Fd())
-
-			for idx, offset := range offsets {
-				if idx%workers != i {
-					continue
-				}
+			filename := fmt.Sprintf("background-write-%d", i)
+			if opts.read {
+				filename = fmt.Sprintf("background-read-%d", i)
+			}
+			filename = filepath.Join(opts.directory, filename)
+			f, err := os.OpenFile(filename, os.O_RDWR|os.O_CREATE|os.O_TRUNC|syscall.O_DIRECT, 0600)
+			if err != nil {
+				log.Printf("Failed to create temp file for background IOPS job: %v", err)
+				return
+			}
 
-				m.Lock()
-				n, err := syscall.Pread(fd, data, offset)
-				m.Unlock()
+			// For O_DIRECT I/O must be aligned on the sector size of the underlying block device.
+			// Usually that's 512 but may also be 4096. Use 4096 since that works in either case.
+			opSize := 4096
+			blocks := 25600
+			fileSize := int64(opSize * blocks)
 
+			if opts.read {
+				_, err := io.Copy(f, io.LimitReader(r, fileSize))
 				if err != nil {
-					errs <- errors.Wrapf(err, "failed to pread %d bytes to %s at offset %d", len(data), filename, offset)
+					log.Printf("Failed to write temp file for background read IOPS jobs: %v", err)
+					return
 				}
-				if n != len(data) {
-					errs <- errors.Wrapf(err, "pread %d of %d bytes to %s at offset %d", n, len(data), filename, offset)
+			} else {
+				err := f.Truncate(int64(opSize * blocks))
+				if err != nil {
+					log.Printf("Failed to resize temp file for backgroupd write IOPS jobs: %v", err)
 				}
 			}
-		}(i)
-	}
 
-	wg.Wait()
-	if len(errs) > 0 {
-		return nil, <-errs
-	}
+			wg.Done()
+			for {
+				if ctx.Err() != nil {
+					if err := os.Remove(filename); err != nil {
+						log.Println(err.Error())
+					}
+					done <- true
+					return
+				}
 
-	d := time.Now().Sub(start)
-	nsPerIO := d / time.Duration(len(offsets))
-	iops := time.Second / nsPerIO
+				mtx.Lock()
+				windowKey := time.Now().Unix()
 
-	fsPerf.IOPS = int(iops)
+				if windows[windowKey] >= opts.iopsLimit {
+					mtx.Unlock()
 
-	collectorName := hostCollector.CollectorName
-	if collectorName == "" {
-		collectorName = "filesystemPerformance"
-	}
-	name := filepath.Join("filesystemPerformance", collectorName+".json")
-	b, err := json.Marshal(fsPerf)
-	if err != nil {
-		return nil, errors.Wrap(err, "failed to marshal fs perf results")
+					nextWindow := windowKey + 1
+					timeUntilNextWindow := time.Until(time.Unix(nextWindow, 0))
+					time.Sleep(timeUntilNextWindow)
+					continue
+				}
+
+				windows[windowKey]++
+				mtx.Unlock()
+
+				blockOffset := rand.Intn(blocks)
+				offset := int64(blockOffset) * int64(opSize)
+				data := make([]byte, opSize)
+
+				if opts.read {
+					syscall.Pread(int(f.Fd()), data, offset)
+				} else {
+					if _, err := rand.Read(data); err != nil {
+						log.Printf("Failed to generate data for background IOPS write job: %v", err)
+						return
```

Comment:
this breaks the 'done' chan count as far as I can tell

Reply:
Wouldn't that make it so that the defer func can never complete if there's an error in this collector somewhere?

Reply:
Yes